src/libraries/System.Private.CoreLib/src/System/Globalization/SurrogateCasing.cs

```diff
@@ -0,0 +1,183 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Runtime.CompilerServices;
+
+namespace System.Globalization
+{
+    internal static class SurrogateCasing
+    {
+        // For simplicity ToUpper doesn't expect the Surrogate be formed with
+        //  S = ((H - 0xD800) * 0x400) + (L - 0xDC00) + 0x10000
+        // Instead it expect to have it in the form (H << 16) | L
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal static void ToUpper(ushort h, ushort l, out ushort hr, out ushort lr)
+        {
+            switch (h)
```

Comment:
This switch statement (and its `ToLower` equivalent) can be deleted. See feedback on _CharUnicodeInfo.cs_ for more info.

However, you'll need to revalidate that _h_ and _l_ are a valid surrogate pair. You can use `Rune.TryCreate` \[non-throwing\] or `new Rune(...)` \[throwing\] as a shortcut for this. The reason you'll need to revalidate this is that the caller may have gotten the original _h_ and _l_ values from a `ReadOnlySpan<char>`, which is __not__ guaranteed to be immutable. So there's a window of opportunity for this mutation:

```cs
ReadOnlySpan<char> span = GetSpan();
if (char.IsHighSurrogate(span[0]) && char.IsLowSurrogate(span[1]))
{
    // Another thread could have updated the contents of 'span' between the check
    // above and the dereference below.
    Consume(span[0], span[1]);
}
```